### PR TITLE
Fix wrong comment about pinning bit and change GetGCSafeMethodTable

### PR DESCRIPTION
**Pull request by 303248153**
Refer to issue #10141
Since pinning bit is moved to sync block,
I change the comment in GetGCSafeMethodTable, and replace the old hard code 3 with MARKED_BIT.

It's a dangerous change, I didn't see any code in gc use the last but one bit, if any of you known please say it.
I also run tests on windows after this change, only one test failed:

```
BEGIN EXECUTION
 "C:\coreclr-fork\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\C
ore_Root\corerun.exe" SizeConstTest.exe

Unhandled Exception: System.Runtime.InteropServices.COMException: The data area passed to a system call is too small (Exception from HRESULT: 0x8007007A)
   at System.StubHelpers.ValueClassMarshaler.ConvertToNative(IntPtr dst, IntPtr
src, IntPtr pMT, CleanupWorkList& pCleanupWorkList)
   at Test.TakeByValTStr(S_CHARArray_ByValTStr s, Int32 size)
   at Test.SizeConstByValTStr()
   at Test.Main(String[] args)
Expected: 100
Actual: -532462766
END EXECUTION - FAILED
FAILED
```

This test also failed before I perform this modification, so I think it's irrelevant.

### CODE REVIEW

File: src/vm/object.h [hunk 1]
```diff
@@ -634,12 +634,12 @@ class Object
...
-        // A method table pointer should always be aligned.  During GC we set the least 
-        // significant bit for marked objects and we set the second to least significant
-        // bit for pinned objects.  So if we want the actual MT pointer during a GC
-        // we must zero out the lowest 2 bits.
-        return dac_cast<PTR_MethodTable>((dac_cast<TADDR>(m_pMethTab)) & ~((UINT_PTR)3));
+        // lose GC marking bit
+        // A method table pointer should always be aligned.
+        // During GC we set the least significant bit for marked objects.
+        // So if we want the actual MT pointer during a GC we must zero out the lowest bit.
+        // In the past we had put pinning bit here but now it's in sync block.
```
Comment (adityamandaleeka) [inline src/vm/object.h]: > In the past we had put pinning bit here but now it's in sync block.
Remove this. It makes it sound like the pinning bit was the lowest bit in the past.

**Comment by adityamandaleeka** [comment]
There's also code in the DAC that's using this technique and should probably be kept in sync.

https://github.com/dotnet/coreclr/blob/master/src/debug/daccess/dacimpl.h#L1682

**Comment by adityamandaleeka** [comment]
@Maoni0 Do you know what the history of this was?

**Comment by adityamandaleeka** [comment]
@303248153 My general advice for testing potentially dangerous changes like this, even if you're entirely sure about the change, is to add a temporary check that verifies (in all builds) that your assumption is true and then run it through all the test CI jobs you can. Once everything is passes, you can remove that check (or make it a normal assert, depending on whether the perf tradeoff is worth it).

In this particular case, you want to make sure that the last two bits are both 0 after you clear the mark bit (take a look at `_ASSERTE_ALL_BUILDS`).

**Comment by Maoni0** [comment]
@adityamandaleeka originally both were on the methodtable*, for server GC the pinned bit was moved to syncblk (for perf); at some point it was decided that there was no need to have this distinction - pinning should be fairly infrequent anyway.

**Comment by 303248153** [comment]
@adityamandaleeka Thanks for your advice,
Looks like the assertion cause a test error, but I can't figure out the what's the reason.
I'll test on my environment and provide more information later.

**Comment by 303248153** [comment]
I run "GC\Regressions\v2.0-rtm\544701\544701\544701.cmd" on my windows 8.1 x64 5 times and they all passed, I saw "Timed Out" in [the error log](https://ci.dot.net/job/dotnet_coreclr/job/master/job/debug_windows_nt_prtest/4681/testReport/junit/GC_Regressions/_v2_0_rtm_544701_544701_544701_/_v2_0_rtm_544701_544701_544701_cmd/), maybe just because assertion cost more times? How can I solve this?

**Comment by adityamandaleeka** [comment]
@303248153 That time-out has nothing to do with your change, and has been observed before.

@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by 303248153** [comment]
Looks like all test passed after I change 3 to 1 and added the assertion, now can I remove the assertion and wait for merge?

**Comment by adityamandaleeka** [comment]
@303248153 Yes, please remove the assertions.

**Comment by 303248153** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test
@dotnet-bot test Tizen armel Cross Release Build

**Comment by adityamandaleeka** [comment]
@303248153 Please address my comment above, and then we can merge your PR.

**Comment by 303248153** [comment]
Ok, I didn't see that before :p

**Comment by jkotas** [comment]
It would be better to just say that the low two bits are used by the GC. 

It would be nice to treat the fact that the GC is using just one of them right now and that it is for marking as GC implementation detail that the VM knows nothing about.

**Comment by 303248153** [comment]
@jkotas
That's reasonable to me, but the comment still should be fixed, I will open another PR and continue to see

**Pull request closed without merging**